src/core/lib/event_engine/windows/windows_engine.cc

```diff
@@ -140,15 +178,171 @@ bool WindowsEventEngine::IsWorkerThread() {
+      on_connect(status);
+    });
+    return invalid_connection_handle;
+  }
+  GRPC_EVENT_ENGINE_TRACE("EventEngine::%p connecting to %s", this,
+                          uri->c_str());
+  // Use dualstack sockets where available.
+  ResolvedAddress address = addr;
+  ResolvedAddress addr6_v4mapped;
+  if (ResolvedAddressToV4Mapped(addr, &addr6_v4mapped)) {
+    address = addr6_v4mapped;
+  }
+  SOCKET sock = WSASocket(AF_INET6, SOCK_STREAM, IPPROTO_TCP, nullptr, 0,
+                          IOCP::GetDefaultSocketFlags());
+  if (sock == INVALID_SOCKET) {
+    Run([on_connect = std::move(on_connect),
+         status = GRPC_WSA_ERROR(WSAGetLastError(), "WSASocket")]() mutable {
+      on_connect(status);
+    });
+    return invalid_connection_handle;
+  }
+  status = PrepareSocket(sock);
+  if (!status.ok()) {
+    Run([on_connect = std::move(on_connect), status]() mutable {
+      on_connect(status);
+    });
+    return invalid_connection_handle;
+  }
+  // Grab the function pointer for ConnectEx for that specific socket It may
+  // change depending on the interface.
+  LPFN_CONNECTEX ConnectEx;
+  GUID guid = WSAID_CONNECTEX;
+  DWORD ioctl_num_bytes;
+  istatus = WSAIoctl(sock, SIO_GET_EXTENSION_FUNCTION_POINTER, &guid,
+                     sizeof(guid), &ConnectEx, sizeof(ConnectEx),
+                     &ioctl_num_bytes, nullptr, nullptr);
+  if (istatus != 0) {
+    Run([on_connect = std::move(on_connect),
+         status = GRPC_WSA_ERROR(
+             WSAGetLastError(),
+             "WSAIoctl(SIO_GET_EXTENSION_FUNCTION_POINTER)")]() mutable {
+      on_connect(status);
+    });
+    return invalid_connection_handle;
+  }
+  // bind the local address
+  auto local_address = ResolvedAddressMakeWild6(0);
+  istatus = bind(sock, local_address.address(), local_address.size());
+  if (istatus != 0) {
+    Run([on_connect = std::move(on_connect),
+         status = GRPC_WSA_ERROR(WSAGetLastError(), "bind")]() mutable {
+      on_connect(status);
+    });
+    return invalid_connection_handle;
+  }
+  // Connect
+  auto watched_socket = iocp_.Watch(sock);
+  auto* info = watched_socket->write_info();
+  bool success =
+      ConnectEx(watched_socket->socket(), address.address(), address.size(),
+                nullptr, 0, nullptr, info->overlapped());
+  // It wouldn't be unusual to get a success immediately. But we'll still get an
+  // IOCP notification, so let's ignore it.
+  if (!success) {
+    int last_error = WSAGetLastError();
+    if (last_error != ERROR_IO_PENDING) {
+      auto status = GRPC_WSA_ERROR(WSAGetLastError(), "ConnectEx");
+      Run([on_connect = std::move(on_connect), status]() mutable {
+        on_connect(status);
+      });
+      watched_socket->MaybeShutdown(status);
+      return invalid_connection_handle;
+    }
+  }
+  GPR_ASSERT(watched_socket != nullptr);
+  auto connection_state = std::make_shared<ConnectionState>();
+  grpc_core::MutexLock lock(&connection_state->mu);
```

Comment:
Is there a need to grab this mutex as soon as connection_state is created ? Can it be moved to line 301 maybe ?

Reply:
`ConnectionState`'s thread guard annotations would be unhappy with that change. We could move some things around, create a constructor for the type instead (which would only partially construct the object), or reduce the set of members protected by the lock, but would it help anything? This lock only affects this one connection, which is going to take just as long whether we lock before or after setting those member variables.